Improve performance of Enum's generic IsDefined / GetName / GetValues / GetNames

stephentoub opened the pull request:
Eliminates the boxing in IsDefined/GetName/GetValues, in GetValues avoids going through the non-generic Array.SetValue for each value, and in GetNames avoids having to go through RuntimeType's GetEnumNames override.

|    Method |           Toolchain |      Mean | Ratio | Allocated |
|---------- |-------------------- |----------:|------:|----------:|
| IsDefined | \master\corerun.exe |  34.39 ns |  1.00 |      24 B |
| IsDefined |     \pr\corerun.exe |  24.37 ns |  0.70 |         - |
|           |                     |           |       |           |
|   GetName | \master\corerun.exe |  52.31 ns |  1.00 |      24 B |
|   GetName |     \pr\corerun.exe |  25.16 ns |  0.48 |         - |
|           |                     |           |       |           |
|  GetNames | \master\corerun.exe |  37.25 ns |  1.00 |      80 B |
|  GetNames |     \pr\corerun.exe |  22.80 ns |  0.61 |      80 B |
|           |                     |           |       |           |
| GetValues | \master\corerun.exe | 573.06 ns |  1.00 |     224 B |
| GetValues |     \pr\corerun.exe |  19.05 ns |  0.03 |      56 B |

```C#
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Diagnosers;
using BenchmarkDotNet.Running;
using System;

[MemoryDiagnoser]
public class Program
{
    static void Main(string[] args) => BenchmarkSwitcher.FromAssemblies(new[] { typeof(Program).Assembly }).Run(args);

    private DayOfWeek _value = DayOfWeek.Friday;

    [Benchmark]
    public bool IsDefined() => Enum.IsDefined(_value);

    [Benchmark]
    public string GetName() => Enum.GetName(_value);

    [Benchmark]
    public string[] GetNames() => Enum.GetNames<DayOfWeek>();

    [Benchmark]
    public DayOfWeek[] GetValues() => Enum.GetValues<DayOfWeek>();
}
```

EgorBo commented on src/libraries/System.Private.CoreLib/src/System/Enum.cs:
```diff
@@ -258,11 +258,27 @@ internal static ulong ToUInt64(object value)
+                TypeCode.UInt64 => Unsafe.As<TEnum, ulong>(ref value),
+                TypeCode.Int64 => (ulong)Unsafe.As<TEnum, long>(ref value),
+                _ => throw new InvalidOperationException(SR.InvalidOperation_UnknownEnumType),
+            };
```
Wonder if this impl is better:
```csharp
private static ulong ToUInt64<TEnum>(TEnum value) where TEnum : struct, Enum
{
    if (!typeof(TEnum).IsPrimitive) // is not currently optimized in JIT but can be.
        throw new InvalidOperationException(SR.InvalidOperation_UnknownEnumType);
    return Unsafe.SizeOf<TEnum>() switch
    {
        8 => Unsafe.As<TEnum, ulong>(ref value),
        4 => Unsafe.As<TEnum, uint>(ref value),
        2 => Unsafe.As<TEnum, ushort>(ref value),
        _ => Unsafe.As<TEnum, byte>(ref value),
    };
}
```

EgorBo replied:
Also, I recall it's actually possible to create a float/double based Enum (and perhaps IntPtr/UIntPtr?)
at least we had a test for it somewhere.

stephentoub replied:
> Also, I recall it's actually possible to create a float/double based Enum (and perhaps IntPtr/UIntPtr?)
at least we had a test for it somewhere.

That already wasn't supported:
https://github.com/dotnet/runtime/blob/640f0f72bc1c31717b6a319ced7e00bd44675349/src/libraries/System.Private.CoreLib/src/System/Enum.cs#L256

> Wonder if this impl is better

IsPrimitive returns true for float/double.

EgorBo replied:
I'll take a look if `Type.GetTypeCode` can be optimized for `typeof()` input into a const.

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Enum.cs:
```diff
@@ -298,7 +314,99 @@ public static Type GetUnderlyingType(Type enumType)
         public static TEnum[] GetValues<TEnum>() where TEnum : struct, Enum
-            => (TEnum[])GetValues(typeof(TEnum));
+        {
+            RuntimeType type = (RuntimeType)typeof(TEnum);
```
This is slow API by design. It allocates a new array on every invocation. Is it really worth it paying trading generic code size expansion to make it faster?

stephentoub replied:
This changed it from being a relatively slow API to a relatively fast API that allocates 1/4th of what it did before. But yes, if anyone were using it on a hot path, we'd recommend they not because of that remaining array allocation, suggesting they do their own caching instead.

My original plan for this PR was just to fix Enum.IsDefined. GetName and GetNames came along for the ride as they were one-line fixes. And then I happened to benchmark GetValues and saw how much waste there was.  If you don't think GetValues is worth it, though, I can revert it.

As a middle-ground, we could also do just:
```C#
public static TEnum[] GetValues<TEnum>() where TEnum : struct, Enum
{
    RuntimeType type = (RuntimeType)typeof(TEnum);

    ulong[] ulValues = InternalGetValues(type);
    var results = new TEnum[ulValues.Length];

    for (int i = 0; i < ulValues.Length; i++)
    {
        results[i] = (TEnum)ToObject(typeof(TEnum), ulValues[i]);
    }

    return results;
}
```
It doesn't eliminate any of the boxing and still pays the cost of the ToObject, but it does avoid going through a non-generic array and most of the additional code size...

|    Method |           Toolchain |      Mean | Ratio | Allocated |
|---------- |-------------------- |----------:|------:|----------:|
| GetValues | \master\corerun.exe | 573.06 ns |  1.00 |     224 B |
| GetValues |     \pr\corerun.exe |  19.05 ns |  0.03 |      56 B |
| GetValues |     \pr2\corerun.exe | 380.31 ns |  0.67 |      224 B |

?

jkotas replied:
My estimate is that the generic expansion will cost about 0.5ms extra in JIT and type loading (assuming non-AOT case). It means that one has to call the API about 1000 times for given `T` to break even. If somebody is calling this API 1000 times for given `T`, they either do not care about performance at all or they should be caching the result.

stephentoub replied:
> My estimate is that the generic expansion will cost about 0.5ms extra in JIT and type loading (assuming non-AOT case)

How did you reach that estimate?  This method is already generic and would be specialized; the difference between:
```C#
(TEnum[])GetValues(typeof(TEnum))
```
and
```C#
public static TEnum[] GetValues<TEnum>() where TEnum : struct, Enum
{
    RuntimeType type = (RuntimeType)typeof(TEnum);

    ulong[] ulValues = InternalGetValues(type);
    var results = new TEnum[ulValues.Length];

    for (int i = 0; i < ulValues.Length; i++)
    {
        results[i] = (TEnum)ToObject(typeof(TEnum), ulValues[i]);
    }

    return results;
}
```
costs ~500us of JIT time? I agree that's not worth it, but it's also way larger than I'd have expected.

jkotas replied:
My estimate was for the original larger method.

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Enum.cs:
```diff
@@ -328,8 +436,8 @@ internal static ulong[] InternalGetValues(RuntimeType enumType)
             RuntimeType enumType = (RuntimeType)typeof(TEnum);
-            ulong[] ulValues = Enum.InternalGetValues(enumType);
-            ulong ulValue = Enum.ToUInt64(value);
+            ulong[] ulValues = InternalGetValues(enumType);
```
It may work better to switch the cached values type to be the underlying enum type. It would eliminate the conversions to UInt64 (we would have to convert or switch over the underlying type instead), and it would also make the cache 2x smaller for the typical case.

`TEnum[] GetValues<TEnum>()` above would then turn into:

```
    Array values = InternalGetValues(type);
    var results = new TEnum[values.Length];
    Array.Copy(values, results, results.Length); // This can call Buffer.MemMove directly if we really cared about squeezing last cycle out of it
    return results;
```

stephentoub replied:
Maybe.  That makes some other APIs more challenging, like InternalFlagsFormat.  I can open an issue to look at doing so, but I'll skip it in this PR.  And since it's complicating things, I'll just revert my GetValues changes.

stephentoub committed:
Improve performance of Enum's generic IsDefined / GetName / GetNames

Eliminates the boxing in IsDefined/GetName/GetValues, and in GetNames avoids having to go through RuntimeType's GetEnumNames override.

EgorBo commented:
@jkotas @stephentoub In case if you are interested: I've just intrinsified `Type.GetTypeCode` for `typeof(x)` or `x.GetType()` args https://github.com/dotnet/runtime/compare/master...EgorBo:opt-type-gettypecode but the jit-diff is not really exciting, just two methods so I guess it's not worth the complexity:
```
(-91.67% of base) : System.Collections.Concurrent.dasm - ConcurrentDictionary`2:IsValueWriteAtomic():bool (7 methods)
(-48.73% of base) : System.Private.CoreLib.dasm - Enum:TryParse(String,bool,bool,byref):bool (6 methods)
```

stephentoub commented:
> so I guess it's not worth the complexity

Probably not right now, but it's good to have in our back pocket for the future :)
